### PR TITLE
Add UID to client-go impersonation config

### PR DESCRIPTION
**Pull request by margocrawf**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Introduces Impersonate-UID to client-go. My previous PR introduced the `Impersonate-UID` header, this PR makes it possible to specify the UID in your impersonation config on the client side in order to pass the `Impersonate-UID` header along with a request.

* Updates ImpersonationConfig in rest/config.go to include UID
  attribute, and pass it through when copying the config
* Updates ImpersonationConfig in transport/config.go to include UID
  attribute
* In transport/round_tripper.go, Set the "Impersonate-Uid" header in
  requests based on the UID value in the config
* Update auth_test.go integration test to specify a UID through the new
  rest.ImpersonationConfig field rather than manually setting the
  Impersonate-Uid header

#### Which issue(s) this PR fixes:

This is another piece of addressing #93699. It is a follow up to #99961. 

#### Does this PR introduce a user-facing change?

```release-note
client-go impersonation config can specify a UID to pass impersonated uid information through in requests.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -207,6 +207,8 @@ type ImpersonationConfig struct {
 	// Extra is a free-form field which can be used to link some authentication information
 	// to authorization information.  This field allows you to impersonate it.
 	Extra map[string][]string
+	// UID is a unique value that identifies the user.
+	UID string
```
Comment (enj) [inline staging/src/k8s.io/client-go/rest/config.go:211]: Put this under `UserName` to be consistent with other places in the codebase.

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -611,6 +613,7 @@ func CopyConfig(config *Config) *Config {
 			Groups:   config.Impersonate.Groups,
 			Extra:    config.Impersonate.Extra,
 			UserName: config.Impersonate.UserName,
+			UID:      config.Impersonate.UID,
```
Comment (enj) [inline staging/src/k8s.io/client-go/rest/config.go:616]: nit: please sort these fields in the order they appear in the struct (after addressing the first comment).

File: staging/src/k8s.io/client-go/rest/transport.go [hunk 3]
```diff
@@ -85,6 +85,7 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 			UserName: c.Impersonate.UserName,
 			Groups:   c.Impersonate.Groups,
 			Extra:    c.Impersonate.Extra,
+			UID:      c.Impersonate.UID,
```
Comment (enj) [inline staging/src/k8s.io/client-go/rest/transport.go:88]: Same comment, move under username.

File: staging/src/k8s.io/client-go/transport/config.go [hunk 4]
```diff
@@ -86,6 +86,8 @@ type ImpersonationConfig struct {
 	Groups []string
 	// Extra matches user.Info.GetExtra()
 	Extra map[string][]string
+	// UID matches user.Info.GetUID()
+	UID string
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/config.go:90]: Same comment, move under username.

File: staging/src/k8s.io/client-go/transport/round_trippers.go [hunk 5]
```diff
@@ -239,6 +243,9 @@ func (rt *impersonatingRoundTripper) RoundTrip(req *http.Request) (*http.Respons
 			req.Header.Add(ImpersonateUserExtraHeaderPrefix+headerKeyEscape(k), v)
 		}
 	}
+	if rt.impersonate.UID != "" {
+		req.Header.Set(ImpersonateUIDHeader, rt.impersonate.UID)
+	}
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/round_trippers.go:248]: nit: move in-between the username and group logic above.

File: staging/src/k8s.io/client-go/transport/round_trippers.go [hunk 6]
```diff
@@ -58,7 +58,8 @@ func HTTPWrappersForConfig(config *Config, rt http.RoundTripper) (http.RoundTrip
 	}
 	if len(config.Impersonate.UserName) > 0 ||
 		len(config.Impersonate.Groups) > 0 ||
-		len(config.Impersonate.Extra) > 0 {
+		len(config.Impersonate.Extra) > 0 ||
+		len(config.Impersonate.UID) > 0 {
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/round_trippers.go:62]: Check this after checking username.

**Review by enj — CHANGES_REQUESTED** [review]
Basically the same comment a bunch of times about field ordering to match the rest of the codebase.

**Comment by deads2k** [comment]
after Mo's comments, this looks plausible to me.  Ping me on slack once updated.

**Pull request merged**